checks/shell_download_validate.go

```diff
@@ -316,6 +323,111 @@ func isExecuteFiles(node syntax.Node, cmd, pathfn string, files map[string]bool,
 	return ok
 }
 
+func isGoUnpinnedDownload(cmd []string) bool {
+	if len(cmd) == 0 {
+		return false
+	}
+
+	if !isBinaryName("go", cmd[0]) {
+		return false
+	}
+
+	// `Go install` will automatically look up the
+	// go.mod and go.sum, so we don't flag it.
+	l := 2
+	if len(cmd) <= l {
+		return false
+	}
+
+	found := false
+	hashRegex := regexp.MustCompile("^[A-Fa-f0-9]{40,}$")
+	for i := 1; i < len(cmd)-1; i++ {
+		// Search for get and install commands.
+		if strings.EqualFold(cmd[i], "install") ||
+			strings.EqualFold(cmd[i], "get") {
+			found = true
+		}
+
+		if !found {
+			continue
+		}
+
+		pkg := cmd[i+1]
+		// Verify pkg = name@hash
+		parts := strings.Split(pkg, "@")
+		if len(parts) != l {
+			continue
+		}
+		hash := parts[1]
+		if hashRegex.Match([]byte(hash)) {
+			return false
+		}
+	}
+
+	return found
+}
+
+func isPipUnpinnedDownload(cmd []string) bool {
+	if len(cmd) == 0 {
+		return false
+	}
+
+	if !isBinaryName("pip", cmd[0]) && !isBinaryName("pip3", cmd[0]) {
+		return false
+	}
+
+	isInstalled := false
+	for i := 1; i < len(cmd)-1; i++ {
+		// Search for get and install commands.
+		if strings.EqualFold(cmd[i], "install") {
+			isInstalled = true
+			continue
+		}
+
+		if !isInstalled || !strings.EqualFold("-r", cmd[i]) {
+			continue
+		}
+
+		requirements := cmd[i+1]
+		if strings.EqualFold(path.Base(requirements), "requirements.txt") {
+			return false
+		}
```

Comment:
what about `pip install package==1.0.0` ?

Reply:
thanks. I will extend the check for `pip install package==hash`